tsk/fs/encryptionHelper.cpp

```diff
@@ -0,0 +1,181 @@
+
+/*
+ ** The Sleuth Kit
+ **
+ ** Brian Carrier [carrier <at> sleuthkit [dot] org]
+ ** Copyright (c) 2024 Sleuth Kit Labs, LLC. All Rights reserved
+ ** Copyright (c) 2010-2021 Brian Carrier.  All Rights reserved
+ **
+ ** This software is distributed under the Common Public License 1.0
+ */
+
+// Methods to handle volume encryption (currently only BitLocker is supported)
+
+#include "encryptionHelper.h"
+
+#ifdef HAVE_LIBMBEDTLS
+#include "tsk/util/Bitlocker/BitlockerParser.h"
+#endif
+
+/**
+* Test whether the volume is encrypted with BitLocker and initialize the parser and other fields if it is.
+* 
+* The theory behind the return values is that we want to get the wrong password / needs password messages back
+* to the user, which means we don't want to overwrite it with any other error codes.
+* 
+* @param a_fs_info  The TSK_FS_INFO object. Should have the img_info and volume offset set but can otherwise be uninitialized.
+*                      Will be updated if we find an successfully initialize BitLocker.
+* @param a_pass     The password or recovery password to use for decryption. May be empty. If the password is not needed
+*                      (for example if we have clear key) it will be ignored.
+* 
+* @return 0 if:
+* - We didn't find the Bitlocker signature
+* - We found encryption and did all the initialization successfully
+* - We found encryption but had an unspecified error in initialization
+* Returns -1 if:
+* - We got far enough to be confident that it's Bitlocker and have a specific error message to get back to the user
+*/
+int handleBitlocker(TSK_FS_INFO* a_fs_info, const char* a_pass) {
+#ifdef HAVE_LIBMBEDTLS
+	BitlockerParser* bitlockerParser = new BitlockerParser();
+	BITLOCKER_STATUS status = bitlockerParser->initialize(a_fs_info->img_info, a_fs_info->offset, a_pass);
+	if (status == BITLOCKER_STATUS::NOT_BITLOCKER) {
+		return 0;
+	}
+
+	if (status != BITLOCKER_STATUS::SUCCESS) {
+
+		// If we have some specific error cases we want to get that information back to the user
+		if (status == BITLOCKER_STATUS::WRONG_PASSWORD) {
```

Comment:
Need to delete bitlockerParser  in all of these conditions

Reply:
I changed the logic a little here so please give it a quick look

Reply:
Looks good.